module/zfs/vdev_removal.c

```diff
@@ -1995,19 +1995,17 @@ spa_vdev_remove_top_check(vdev_t *vd)
 		if (available < vd->vdev_stat.vs_alloc)
 			return (SET_ERROR(ENOSPC));
 	} else {
-		/* available space in the pool's normal class */
+		/*
+		 * This is a normal device. There has to be enough free
+		 * space in the pool's normal class to remove the device
+		 * and leave 3% of the space available.  The available
+		 * space accounts for both reserved and "slop" space.
+		 */
 		uint64_t available = dsl_dir_space_available(
 		    spa->spa_dsl_pool->dp_root_dir, NULL, 0, B_TRUE);
-		if (available <
-		    vd->vdev_stat.vs_dspace + spa_get_slop_space(spa)) {
-			/*
-			 * This is a normal device. There has to be enough free
-			 * space to remove the device and leave double the
-			 * "slop" space (i.e. we must leave at least 3% of the
-			 * pool free, in addition to the normal slop space).
-			 */
+
+		if (available - (available >> 5) < vd->vdev_stat.vs_dspace)
```

Comment:
I don't think this is calculating what the comment says.  "leave 3% of the space available" to me means that after the removal, AVAIL (dsl_dir_space_available(root_dir)) should be at least 3% (1/2^5) of the size of the pool after the removal.

For example, imagine we have a 100TB pool, are removing a 1TB device, and have 1.04TB AVAIL.  The code here would allow the removal to proceed (1.04-1.04>>5=1.0075TB).  After the removal there would be 0.0075TB AVAIL, but 3% of 99TB=~3TB.

Reply:
Indeed.  I've updated the calculation accordingly.   It now checks that the available space after the removal isn't less than `spa_get_dspace() / (1^6)`.  i.e. we always leave at least 1.6% of the pools normal class available.  So for your example.

1.04TB AVAIL - 1.00TB = 0.04TB after remove which is less than 100TB / (1^6) = 1.6TB.